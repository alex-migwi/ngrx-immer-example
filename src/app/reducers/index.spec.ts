import { getProducts, getProductSkus, getCatalog, getCartItems, getAllCartSummary, getCartSummary, State } from './';
import { Product, CartItem } from '../models/index';

const createProduct = ({ sku = '', name = '', image = '', price = 1 } = {}): Product => ({
  sku: sku,
  name: name || `name-${sku}`,
  price,
  image: image || `image-${sku}`,
});

const createState = ({
  catalog = {
    products: {
      'PRODUCT-AAA': createProduct({ sku: 'PRODUCT-AAA' }),
      'PRODUCT-BBB': createProduct({ sku: 'PRODUCT-BBB' }),
      'PRODUCT-CCC': createProduct({ sku: 'PRODUCT-CCC' }),
    },
    productSkus: ['PRODUCT-AAA', 'PRODUCT-BBB', 'PRODUCT-CCC'],
  },
  cart = {
    cartItems: {
      'PRODUCT-AAA': 3,
      'PRODUCT-CCC': 0,
    },
  },
} = {}): State => ({
  catalog,
  cart,
});

// step 1
test('getProducts', () => {
  expect(Object.keys(getProducts(createState())).length).toEqual(3);
});

test('getProductSkus', () => {
  expect(getProductSkus(createState()).length).toBe(3);
});

test('getCatalog', () => {
  expect(getCatalog(createState()).length).toBe(3);
});

test('getCartItems', () => {
  expect(Object.keys(getCartItems(createState())).length).toBe(2);
});

test('getAllCartSummary', () => {
  expect(getAllCartSummary(createState()).length).toBe(2);
});

test('getCartSummary', () => {
  expect(getCartSummary(createState()).length).toBe(1);
});

// step 2
const fixtures = [
  {
    name: 'getProducts',
    selector: getProducts,
    state: createState(),
  },
  {
    name: 'getProductSkus',
    selector: getProductSkus,
    state: createState(),
  },
  {
    name: 'getCatalog',
    selector: getCatalog,
    state: createState(),
  },
  {
    name: 'getCartItems',
    selector: getCartItems,
    state: createState(),
  },
  {
    name: 'getAllCartSummary',
    selector: getAllCartSummary,
    state: createState(),
  },
  {
    name: 'getCartSummary',
    selector: getCartSummary,
    state: createState(),
  },
];

fixtures.forEach(fixture => {
  test(fixture.name, () => {
    expect(fixture.selector(fixture.state)).toMatchSnapshot();
  });
});

// step 3
test('getCartSummary with projector', () => {
  const cartItems: CartItem[] = [
    {
      amount: 1,
      product: createProduct(),
    },
    {
      amount: 0,
      product: createProduct(),
    },
    {
      amount: 2,
      product: createProduct(),
    },
  ];

  expect(getCartSummary.projector(cartItems).length).toBe(2);
});
