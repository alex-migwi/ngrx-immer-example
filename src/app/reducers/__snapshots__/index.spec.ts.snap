// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllCartSummary 1`] = `
Array [
  Object {
    "amount": 3,
    "product": Object {
      "image": "image-PRODUCT-AAA",
      "name": "name-PRODUCT-AAA",
      "price": 1,
      "sku": "PRODUCT-AAA",
    },
  },
  Object {
    "amount": 0,
    "product": Object {
      "image": "image-PRODUCT-CCC",
      "name": "name-PRODUCT-CCC",
      "price": 1,
      "sku": "PRODUCT-CCC",
    },
  },
]
`;

exports[`getCartItems 1`] = `
Object {
  "PRODUCT-AAA": 3,
  "PRODUCT-CCC": 0,
}
`;

exports[`getCartSummary 1`] = `
Array [
  Object {
    "amount": 3,
    "product": Object {
      "image": "image-PRODUCT-AAA",
      "name": "name-PRODUCT-AAA",
      "price": 1,
      "sku": "PRODUCT-AAA",
    },
  },
]
`;

exports[`getCatalog 1`] = `
Array [
  Object {
    "image": "image-PRODUCT-AAA",
    "name": "name-PRODUCT-AAA",
    "price": 1,
    "sku": "PRODUCT-AAA",
  },
  Object {
    "image": "image-PRODUCT-BBB",
    "name": "name-PRODUCT-BBB",
    "price": 1,
    "sku": "PRODUCT-BBB",
  },
  Object {
    "image": "image-PRODUCT-CCC",
    "name": "name-PRODUCT-CCC",
    "price": 1,
    "sku": "PRODUCT-CCC",
  },
]
`;

exports[`getProductSkus 1`] = `
Array [
  "PRODUCT-AAA",
  "PRODUCT-BBB",
  "PRODUCT-CCC",
]
`;

exports[`getProducts 1`] = `
Object {
  "PRODUCT-AAA": Object {
    "image": "image-PRODUCT-AAA",
    "name": "name-PRODUCT-AAA",
    "price": 1,
    "sku": "PRODUCT-AAA",
  },
  "PRODUCT-BBB": Object {
    "image": "image-PRODUCT-BBB",
    "name": "name-PRODUCT-BBB",
    "price": 1,
    "sku": "PRODUCT-BBB",
  },
  "PRODUCT-CCC": Object {
    "image": "image-PRODUCT-CCC",
    "name": "name-PRODUCT-CCC",
    "price": 1,
    "sku": "PRODUCT-CCC",
  },
}
`;
